# Taller 12 Optimización y analitica dualidad

por: Catalina Castro, Gabriella Espitia, Isabella Parales, Juan Diego Saldarriaga, Nestor Tabares

## Punto 0

In [1]:
from pulp import *

In [2]:

# Variables de decisión
x_i1 = LpVariable("x_i1", lowBound=0)  # Variable para producto i1 (tablero)
x_i2 = LpVariable("x_i2", lowBound=0)  # Variable para producto i2 (cajas)

# Problema PL
model_punto0 = LpProblem("Maximizar_Beneficio", LpMaximize)

# Función objetivo
model_punto0 += 100000 * x_i1 + 50000 * x_i2

# Restricción de capacidad de almacenamiento
R1 = 20 * x_i1 + 10 * x_i2 <= 2000
model_punto0 += R1

# Restricción de permitido de contaminación
R2 = 0.02 * x_i1 + 0.03 * x_i2 <= 1
model_punto0 += R2

# Resolver el modelo
model_punto0.solve()

# Imprimir el estado de la solución
print("Estado:", LpStatus[model_punto0.status])

print(f"x_i1 (tablero) = {x_i1.varValue}")
print(f"x_i2 (cajas) = {x_i2.varValue}")

print("Valor de la función objetivo:", value(model_punto0.objective))

print("\nPrecios sombra (valores marginales):")
print(f"Capacidad de Almacenamiento (R1) = {R1.pi}")
print(f"Permitido de Contaminación (R2) = {R2.pi}")


Estado: Optimal
x_i1 (tablero) = 50.0
x_i2 (cajas) = 0.0
Valor de la función objetivo: 5000000.0

Precios sombra (valores marginales):
Capacidad de Almacenamiento (R1) = -0.0
Permitido de Contaminación (R2) = 5000000.0


El gerente de PDP_SA deberia considerar negociar una oferta formal respecto al almacen de la fabrica adyacente por un costo de 700 u.m , dado que el costo es más grande que el marginal de almacenamiento(R1)

In [3]:
# Opción 1
diferencia = 0.9 -1
costo = R2.pi * diferencia
print("Opción 1")
print(f"La maxima ganancia teniendo en cuenta la nueva normativa de la maxima emisión del 0.9 es de {value(model_punto0.objective)+costo} ")
print("---"*100)
print("Opción 2")
print(f"La maxima ganancia al asumir la multa de 70000 u.m por exceder la maxima emisión permitido es de {value(model_punto0.objective)-70000}")
print("---"*100)
if value(model_punto0.objective)+ costo> value(model_punto0.objective)-70000:
    print(" La opción 1 es la mejor")
else: 
    print("La opción 2 es mejor ")
    


Opción 1
La maxima ganancia teniendo en cuenta la nueva normativa de la maxima emisión del 0.9 es de 4500000.0 
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Opción 2
La maxima ganancia al asumir la multa de 70000 u.m por exceder la maxima emisión permitido es de 4930000.0
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
La opción 2 es mejor 


Pagar la multa no es ético porque daña el medio ambiente, ignora la responsabilidad social y legal, y compromete la sostenibilidad a largo plazo.

## Punto 1

In [4]:
# Variables de decisión
w_i1 = LpVariable("w_i1", lowBound=0)  # Variable para horas tipo A
w_i2 = LpVariable("w_i2", lowBound=0)  # Variable para horas tipo B



# Problema PL
model_punto1 = LpProblem("Maximizar_Beneficio", LpMaximize)
# Función objetivo
model_punto1 += 20 * w_i1 + 40 * w_i2, "Función_Objetivo"

# Restricción de capacidad de horas hombre
model_punto1 += 30 * w_i1 + 20 * w_i2 <= 2700, "Capacidad_de_Horas_Hombre"

# Restricción de capacidad de hombres máquina
model_punto1 += 5 * w_i1 + 10 * w_i2 <= 850, "Capacidad_de_Hombres_Máquina"

# Restricción de demanda mínima
model_punto1 += w_i1 + w_i2 >= 95, "Demanda"

# Resolver el modelo
model_punto1.solve()

# Imprimir el estado de la solución
print("Estado:", LpStatus[model_punto1.status])

# Imprimir la solución óptima
print(f"w_i1 (horas tipo A) = {w_i1.varValue}")
print(f"w_i2 (horas tipo B) = {w_i2.varValue}")

# Imprimir el valor de la función objetivo
print("Valor de la función objetivo:", value(model_punto1.objective))

print("\nPrecios sombra (valores marginales):")
for nombre, restriccion in model_punto1.constraints.items():
    print(f"{nombre}: = {restriccion.pi}")


print("print")


Estado: Optimal
w_i1 (horas tipo A) = 20.0
w_i2 (horas tipo B) = 75.0
Valor de la función objetivo: 3400.0

Precios sombra (valores marginales):
Capacidad_de_Horas_Hombre: = -0.0
Capacidad_de_Hombres_Máquina: = 4.0
Demanda: = -0.0
print


La inversión de la empresa es incrorecta porque no maximiza los beneficios
de la empresa pues las horas de ingeniero no dan mayor utilidad mientras
que las de máquina si. Cada hora de máquina más da $4 más, en cambio las
horas ingenieron no aumentan beneficio.

La empresa podría reducir su acuerdo con el cliente porque no perdería dinero
pero no es que los beneficios vayan a aumentar, se van a mantener.

## Punto 2

In [5]:

def punto2 ( hh = 56):
    # Variables de decisión
    y_i1 = LpVariable("y_i1", lowBound=0)  # Variable para el producto i1 (pasteles)
    y_i2 = LpVariable("y_i2", lowBound=0)  # Variable para el producto i2 (galletas)

    # Problema PL
    model_punto2 = LpProblem("Maximizar_Beneficio", LpMaximize)
    # Función objetivo
    model_punto2 += 70 * y_i1 + 90 * y_i2

    # Restricción de horas de preparación
    R12 = 4 * y_i1 + 3 * y_i2 <= 40
    model_punto2 += R12

    # Restricción de horas de horneado
    R22 = 4 * y_i1 + 7 * y_i2 <= hh
    model_punto2 += R22

    # Resolver el modelo
    model_punto2.solve()

    # Imprimir el estado de la solución
    print("Estado:", LpStatus[model_punto2.status])

    print(f"y_i1 (pasteles) = {y_i1.varValue}")
    print(f"y_i2 (galletas) = {y_i2.varValue}")

    print("Valor de la función objetivo:", value(model_punto2.objective))

    print("\nPrecios sombra (valores marginales):")
    print(f"Maximo de horas de preparación (R1) = {R12.pi}")
    print(f"Maximo de horas de horneado (R2) = {R22.pi}")


In [6]:
print("Con 56 horas de horneado diario")
punto2()
print("---"*100)
print("Con 72 horas de horneado diario")
punto2(hh = 72)

Con 56 horas de horneado diario
Estado: Optimal
y_i1 (pasteles) = 7.0
y_i2 (galletas) = 4.0
Valor de la función objetivo: 850.0

Precios sombra (valores marginales):
Maximo de horas de preparación (R1) = 8.125
Maximo de horas de horneado (R2) = 9.375
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Con 72 horas de horneado diario
Estado: Optimal
y_i1 (pasteles) = 4.0
y_i2 (galletas) = 8.0
Valor de la función objetivo: 1000.0

Precios sombra (valores marginales):
Maximo de horas de preparación (R1) = 8.125
Maximo de horas de horneado (R2) = 9.375


El marginal arroja un valor de 9.375, por lo cual el aumento de 16 horas daría un aumento en las
ganancias de 150, aunque las horas de preparación son un limitante, al aumenar las horas de
horneado la organización de cantidad de producto a producir cambia, por lo cual al final
este aumento de horas si genera una ganancia a la panadería

## Punto 3

In [7]:
def punto3(qa = 80000):
    # Variables de decisión
    z_i1 = LpVariable("z_i1", lowBound=0)  # Variable para químico A
    z_i2 = LpVariable("z_i2", lowBound=0)  # Variable para químico S1A
    z_i3 = LpVariable("z_i3", lowBound=0)  # Variable para químico S2A

    # Crear el modelo de optimización
    model_punto3 = LpProblem("Maximizar_Beneficio", LpMaximize)
    # Función objetivo
    model_punto3 += 200 * z_i1 + 300 * z_i2 + 320 * z_i3, "Función_Objetivo"

    # Restricción de almacenamiento
    model_punto3 += z_i1 + z_i2 + z_i3 <= 100000, "Almacenamiento"

    # Restricción de empaques de químico A
    model_punto3 += z_i1 >= 60000, "Empaque_Quimico_A"

    # Restricción de capacidad de producción para S2A
    model_punto3 += z_i3 <= 12000, "Capacidad_Producción_S2A"

    # Restricción de capacidad total de producción
    model_punto3 += z_i1 + z_i2 + z_i3 <= qa, "Capacidad_Producción_Total"

    # Restricción de capacidad de producción para S1A
    model_punto3 += z_i2 <= 40000, "Capacidad_Producción_S1A"

    # Restricción de relación entre S1A y S2A
    model_punto3 += 3 * z_i2 == z_i3, "Relacion_S1A_S2A"

    # Resolver el modelo
    model_punto3.solve()

    # Imprimir el estado de la solución
    print("Estado:", LpStatus[model_punto3.status])

    print(f"z_i1 (químico A) = {z_i1.varValue}")
    print(f"z_i2 (químico S1A) = {z_i2.varValue}")
    print(f"z_i3 (químico S2A) = {z_i3.varValue}")

    print("Valor de la función objetivo:", value(model_punto3.objective))

    print("\nPrecios sombra (valores marginales):")
    for nombre, restriccion in model_punto3.constraints.items():
        print(f"{nombre}: {restriccion.pi}")

In [8]:
print("Administrar el quimico A a 80000 litros")
punto3()
print("---"*100)
print("Administrar el equipo A a 75000 litros")
punto3(qa = 75000)

Administrar el quimico A a 80000 litros
Estado: Optimal
z_i1 (químico A) = 64000.0
z_i2 (químico S1A) = 4000.0
z_i3 (químico S2A) = 12000.0
Valor de la función objetivo: 17840000.0

Precios sombra (valores marginales):
Almacenamiento: -0.0
Empaque_Quimico_A: -0.0
Capacidad_Producción_S2A: 153.33333
Capacidad_Producción_Total: 200.0
Capacidad_Producción_S1A: -0.0
Relacion_S1A_S2A: 33.333333
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Administrar el equipo A a 75000 litros
Estado: Optimal
z_i1 (químico A) = 60000.0
z_i2 (químico S1A) = 3750.0
z_i3 (químico S2A) = 11250.0
Valor de la función objetivo: 16725000.0

Precios sombra (valores marginales):
Almacenamiento: -0.0
Empaque_Quimico_A: -115.0
Capacidad_Producción_S2A: -0.0
Capacidad_

Si se administran 75,000 litros del químico A, la ganancia máxima se reducirá en 1,115,000 unidades monetarias. Además, la producción de todos los productos del tipo i disminuirá.